src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,151 @@
+use indexmap::IndexSet;
+
 use crate::term::{Book, Definition, Name, Rule, Term};
-use std::{
-  collections::{BTreeMap, BTreeSet},
-  ops::BitAnd,
-};
+use std::collections::BTreeMap;
+
+type Combinators = BTreeMap<Name, Definition>;
 
-/// Replaces closed Terms (i.e. without free variables) with a Ref to the extracted term
-/// Precondition: Vars must have been sanitized
 impl Book {
+  /// Extracts unsafe terms into new definitions.
+  ///
+  /// Precondition: Variables must have been sanitized.
+  ///
+  /// The floating algorithm follows these rules:
+  /// - Don't extract safe terms or terms that contains unscoped variables.
+  /// - Extract if it is a closed Application
+  /// - Extract if it is a supercombinator
+  /// - Float every element of a Superposition
   pub fn float_combinators(&mut self) {
     let mut combinators = Combinators::new();
 
+    let slf = self.clone();
     for (def_name, def) in self.defs.iter_mut() {
+      let mut name_gen = 0;
+
       if self.entrypoint.as_ref().is_some_and(|m| m == def_name) {
         continue;
       }
 
       let builtin = def.builtin;
       let rule = def.rule_mut();
-      rule.body.float_combinators(def_name, builtin, &mut combinators);
+      let mut seen = IndexSet::new();
+      rule.body.float_combinators(&mut combinators, &mut name_gen, &slf, def_name, builtin, &mut seen);
     }
 
-    // Definitions are not inserted to the book as they are defined to appease the borrow checker.
-    // Since we are mut borrowing the rules we can't borrow the book to insert at the same time.
     self.defs.extend(combinators);
   }
 }
 
-type Combinators = BTreeMap<Name, Definition>;
-
-#[derive(Debug)]
-struct TermInfo<'d> {
-  // Number of times a Term has been detached from the current Term
-  counter: u32,
-  def_name: Name,
-
-  builtin: bool,
-  needed_names: BTreeSet<Name>,
-  combinators: &'d mut Combinators,
-}
-
-impl<'d> TermInfo<'d> {
-  fn new(def_name: Name, builtin: bool, combinators: &'d mut Combinators) -> Self {
-    Self { counter: 0, def_name, builtin, needed_names: BTreeSet::new(), combinators }
-  }
-  fn request_name(&mut self, name: &Name) {
-    self.needed_names.insert(name.clone());
-  }
+impl Term {
+  fn float_combinators(
+    &mut self,
+    combinators: &mut Combinators,
+    name_gen: &mut usize,
+    book: &Book,
+    def_name: &Name,
+    builtin: bool,
+    seen: &mut IndexSet<Name>,
+  ) {
+    for term in self.children_mut() {
+      // Recursively float the children terms.
+      term.float_combinators(combinators, name_gen, book, def_name, builtin, seen);
+
+      // Don't float if it has unscoped variables.
+      if term.has_unscoped_diff() {
+        return;
```

Comment:
Don't return here but rather `continue`, since the siblings might be fine to float.

Reply:
Alternatively, add this condition to the `if` below or to the `is_combinator` function (since it kinda captures the idea of an unscoped combinator)